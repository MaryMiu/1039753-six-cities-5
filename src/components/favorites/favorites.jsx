import React, {PureComponent} from "react";
import PropTypes from "prop-types";
import Header from "../header/header";
import {connect} from "react-redux";
import FavoritesLocation from "../favorites-location/favorites-location";
import FavoritesEmpty from "../favorites-empty/favorites-empty";
import {fetchFavoriteList} from "../../store/api-actions";
import {getFavoriteOffers, getFavoriteIds} from "../../store/selectors";

class Favorites extends PureComponent {
  constructor(props) {
    super(props);
  }

  componentDidMount() {
    const {fetchFavoriteListAction} = this.props;
    fetchFavoriteListAction();
  }

  render() {
    const {favoritesOffers, favoriteIds} = this.props;
    const filtredFavoritesOffers = favoritesOffers.filter((favoritesOffer) => favoriteIds.includes(favoritesOffer.id));
    const favoriteCities = filtredFavoritesOffers.map((favoritesOffer) => favoritesOffer.city.name);
    const uniqueCities = Array.from(new Set(favoriteCities));
    return (
      <div className={`page ${filtredFavoritesOffers.length === 0 ? `page--favorites-empty` : ``}`}>
        <Header />
        {filtredFavoritesOffers.length > 0 ?
          <main className="page__main page__main--favorites">
            <div className="page__favorites-container container">
              <section className="favorites">
                <h1 className="favorites__title">Saved listing</h1>
                <ul className="favorites__list">
                  {uniqueCities.map((uniqueCity) => {
                    return (
                      <FavoritesLocation key={uniqueCity} city={uniqueCity} offers={filtredFavoritesOffers.filter((filtredFavoriteOffer) => filtredFavoriteOffer.city.name === uniqueCity)} />
                    );
                  })}
                </ul>
              </section>
            </div>
          </main>
          :
          <FavoritesEmpty />
        }
        <footer className="footer container">
          <a className="footer__logo-link" href="main.html">
            <img className="footer__logo" src="/img/logo.svg" alt="6 cities logo" width="64" height="33" />
          </a>
        </footer>
      </div>
    );
  }
}

Favorites.propTypes = {
  favoritesOffers: PropTypes.array.isRequired,
  favoriteIds: PropTypes.array.isRequired,
  fetchFavoriteListAction: PropTypes.func.isRequired,
};

const mapStateToProps = (state) => ({
  favoritesOffers: getFavoriteOffers(state),
  favoriteIds: getFavoriteIds(state),
});

const mapDispatchToProps = (dispatch) => {
  return {
    fetchFavoriteListAction: () => dispatch(fetchFavoriteList())
  };
};

export {Favorites};
export default connect(mapStateToProps, mapDispatchToProps)(Favorites);
