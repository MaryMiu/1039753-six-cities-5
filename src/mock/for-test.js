export const offers = [{
  city: {
    name: `Cologne`,
    location: {
      latitude: 50.938361,
      longitude: 6.959974,
      zoom: 13
    }
  },
  images: [
    `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/15.jpg`,
    `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/16.jpg`,
    `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/9.jpg`,
    `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/10.jpg`,
    `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/4.jpg`,
    `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/19.jpg`,
    `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/14.jpg`,
    `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/3.jpg`,
    `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/2.jpg`,
    `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/11.jpg`,
    `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/1.jpg`,
    `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/17.jpg`,
    `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/5.jpg`,
    `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/13.jpg`
  ],
  title: `The house among olive `,
  rating: 3.8,
  type: `HOUSE`,
  bedrooms: 3,
  price: 493,
  goods: [
    `Laptop friendly workspace`,
    `Washer`,
    `Towels`,
    `Fridge`,
    `Air conditioning`,
    `Baby seat`,
    `Breakfast`
  ],
  host: {
    avatarUrl: `img/avatar-angelina.jpg`,
    isPro: true
  },
  description: `Discover daily local life in city center, friendly neighborhood, clandestine casino, karaoke, old-style artisans, art gallery and artist studio downstairs.`,
  location: {
    latitude: 50.957361,
    longitude: 6.9509739999999995,
    zoom: 16
  },
  id: 1,
  isPremium: false,
  isFavorite: false,
  maxAdults: 4,
  previewImage: `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/14.jpg`
},
{
  city: {
    name: `Brussels`,
    location: {
      latitude: 50.846557,
      longitude: 4.351697,
      zoom: 13
    }
  },
  images: [
    `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/3.jpg`,
    `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/10.jpg`,
    `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/11.jpg`,
    `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/20.jpg`,
    `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/14.jpg`,
    `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/7.jpg`,
    `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/13.jpg`,
    `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/19.jpg`,
    `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/18.jpg`,
    `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/16.jpg`,
    `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/8.jpg`,
    `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/15.jpg`,
    `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/9.jpg`,
    `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/4.jpg`
  ],
  title: `Amazing and Extremely Central Flat`,
  rating: 4.3,
  type: `APARTMENT`,
  bedrooms: 3,
  price: 323,
  goods: [
    `Towels`,
    `Baby seat`,
    `Laptop friendly workspace`,
    `Dishwasher`,
    `Washing machine`,
    `Air conditioning`,
    `Washer`,
    `Coffee machine`,
    `Breakfast`,
    `Fridge`
  ],
  host: {
    avatarUrl: `img/avatar-angelina.jpg`,
    isPro: true
  },
  description: `This is a place for dreamers to reset, reflect, and create. Designed with a \`slow\` pace in mind, our hope is that you enjoy every part of your stay; from making local coffee by drip in the morning, choosing the perfect record to put on as the sun sets.`,
  location: {
    latitude: 50.839557,
    longitude: 4.346697,
    zoom: 16
  },
  id: 2,
  isPremium: false,
  isFavorite: false,
  maxAdults: 8,
  previewImage: `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/8.jpg`
},
{
  city: {
    name: `Brussels`,
    location: {
      latitude: 50.846557,
      longitude: 4.351697,
      zoom: 13
    }
  },
  images: [
    `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/2.jpg`,
    `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/1.jpg`,
    `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/16.jpg`,
    `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/12.jpg`,
    `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/17.jpg`,
    `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/5.jpg`,
    `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/11.jpg`,
    `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/6.jpg`,
    `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/4.jpg`,
    `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/8.jpg`,
    `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/19.jpg`,
    `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/7.jpg`,
    `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/10.jpg`,
    `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/3.jpg`
  ],
  title: `The house among olive `,
  rating: 4.6,
  type: `APARTMENT`,
  bedrooms: 4,
  price: 218,
  goods: [
    `Washer`,
    `Laptop friendly workspace`,
    `Breakfast`,
    `Air conditioning`,
    `Baby seat`
  ],
  host: {
    avatarUrl: `img/avatar-angelina.jpg`,
    isPro: true
  },
  description: `Relax, rejuvenate and unplug in this ultimate rustic getaway experience in the country. In our beautiful screened Pondhouse, you can gaze at the stars and listen to the sounds of nature from your cozy warm bed.`,
  location: {
    latitude: 50.865556999999995,
    longitude: 4.371696999999999,
    zoom: 16
  },
  id: 3,
  isPremium: false,
  isFavorite: false,
  maxAdults: 7,
  previewImage: `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/16.jpg`
}
];

export const offer = {
  city: {
    name: `Cologne`,
    location: {
      latitude: 50.938361,
      longitude: 6.959974,
      zoom: 13
    }
  },
  images: [
    `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/15.jpg`,
    `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/16.jpg`,
    `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/9.jpg`,
    `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/10.jpg`,
    `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/4.jpg`,
    `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/19.jpg`,
    `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/14.jpg`,
    `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/3.jpg`,
    `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/2.jpg`,
    `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/11.jpg`,
    `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/1.jpg`,
    `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/17.jpg`,
    `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/5.jpg`,
    `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/13.jpg`
  ],
  title: `The house among olive `,
  rating: 3.8,
  type: `HOUSE`,
  bedrooms: 3,
  price: 493,
  goods: [
    `Laptop friendly workspace`,
    `Washer`,
    `Towels`,
    `Fridge`,
    `Air conditioning`,
    `Baby seat`,
    `Breakfast`
  ],
  host: {
    avatarUrl: `img/avatar-angelina.jpg`,
    isPro: true
  },
  description: `Discover daily local life in city center, friendly neighborhood, clandestine casino, karaoke, old-style artisans, art gallery and artist studio downstairs.`,
  location: {
    latitude: 50.957361,
    longitude: 6.9509739999999995,
    zoom: 16
  },
  id: 1,
  isPremium: false,
  isFavorite: false,
  maxAdults: 4,
  previewImage: `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/14.jpg`
};

export const favoriteIds = [13, 15, 24, 26];

export const AuthorizationStatus = {
  AUTH: `AUTH`,
  NO_AUTH: `NO_AUTH`,
};

export const activeCity = `Paris`;

export const activeSortType = `Popular`;

export const review = {
  id: 1,
  user: {
    avatarUrl: `https://assets.htmlacademy.ru/intensives/javascript-3/avatar/6.jpg`,
    isPro: false
  },
  rating: 2,
  comment: `The room was spacious and clean. The pool looked nothing like the photos and desparately needs a clean. The sauna and spa were closed for lunar new year holiday.`,
  date: `2020-10-28T13:38:44.753Z`
};

export const reviews = [{
  id: 1,
  user: {
    avatarUrl: `https://assets.htmlacademy.ru/intensives/javascript-3/avatar/6.jpg`,
    isPro: false
  },
  rating: 2,
  comment: `The room was spacious and clean. The pool looked nothing like the photos and desparately needs a clean. The sauna and spa were closed for lunar new year holiday.`,
  date: `2020-10-28T13:38:44.753Z`
},
{
  id: 2,
  user: {
    avatarUrl: `https://assets.htmlacademy.ru/intensives/javascript-3/avatar/6.jpg`,
    isPro: false
  },
  rating: 3,
  comment: `What an amazing view! The house is stunning and in an amazing location. The large glass wall had an amazing view of the river!`,
  date: `2020-10-28T13:38:44.753Z`
},
{
  id: 3,
  user: {
    avatarUrl: `https://assets.htmlacademy.ru/intensives/javascript-3/avatar/2.jpg`,
    isPro: false
  },
  rating: 4,
  comment: `The deluxe room was a quite comfortable one with all the adequate facilities. The only thing that made me feel uncomfortable was the rude behavior of an impolite staff at the reception desk.`,
  date: `2020-10-28T13:38:44.753Z`
}
];

export const currentClasses = {
  listClass: `near-places__list`,
  cardClass: `near-places__card`,
  imgClass: `near-places__image-wrapper`
};

export const match = {
  path: `/offer/:id`,
  url: `/offer/11`,
  isExact: true,
  params: {
    id: `11`
  }
};

export const mapStyle = {
  display: `flex`,
  height: `100%`,
  width: 1144,
  margin: `0 auto`
};

export const offersId = [1, 2, 3];
