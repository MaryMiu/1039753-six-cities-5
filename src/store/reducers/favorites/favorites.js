import {extend} from "../../../utils";
import ActionType from "../../constants";

const initialState = {
  favoriteIds: [],
};

const favorites = (state = initialState, action) => {
  switch (action.type) {
    case ActionType.ADD_FAVORITE:
      return extend(state, {
        favoriteIds: [...state.favoriteIds, action.payload],
      });
    case ActionType.DELETE_FAVORITE:
      return extend(state, {
        favoriteIds: state.favoriteIds.filter((id) => action.payload !== id),
      });
    case ActionType.UPDATE_FAVORITES:
      return extend(state, {
        favoriteIds: action.payload,
      });
  }

  return state;
};

export {favorites};
