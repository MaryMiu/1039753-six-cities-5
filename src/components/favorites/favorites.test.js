import React from "react";
import renderer from "react-test-renderer";
import {Favorites} from "./favorites";
import {offers} from "../../mock/for-test";
import {favoriteIds} from "../../mock/for-test";

const noop = () => {};

jest.mock(`../header/header`, () => `Header`);
jest.mock(`../favorites-location/favorites-location`, () => `FavoritesLocation`);
jest.mock(`../favorites-empty/favorites-empty`, () => `FavoritesEmpty`);

describe(`Header render correctly`, () => {

  it(`Should Favorites render correctly with offers`, () => {
    const tree = renderer
    .create(<Favorites
      favoritesOffers={offers}
      favoriteIds={favoriteIds}
      fetchFavoriteListAction={noop}
    />)
    .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it(`Should Favorites render correctly without offers`, () => {
    const tree = renderer
    .create(<Favorites
      favoritesOffers={[]}
      favoriteIds={favoriteIds}
      fetchFavoriteListAction={noop}
    />)
    .toJSON();

    expect(tree).toMatchSnapshot();
  });

});
