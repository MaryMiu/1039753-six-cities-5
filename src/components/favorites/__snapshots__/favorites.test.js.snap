// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header render correctly Should Favorites render correctly with offers 1`] = `
<div
  className="page page--favorites-empty"
>
  <Header />
  <FavoritesEmpty />
  <footer
    className="footer container"
  >
    <a
      className="footer__logo-link"
      href="main.html"
    >
      <img
        alt="6 cities logo"
        className="footer__logo"
        height="33"
        src="/img/logo.svg"
        width="64"
      />
    </a>
  </footer>
</div>
`;

exports[`Header render correctly Should Favorites render correctly without offers 1`] = `
<div
  className="page page--favorites-empty"
>
  <Header />
  <FavoritesEmpty />
  <footer
    className="footer container"
  >
    <a
      className="footer__logo-link"
      href="main.html"
    >
      <img
        alt="6 cities logo"
        className="footer__logo"
        height="33"
        src="/img/logo.svg"
        width="64"
      />
    </a>
  </footer>
</div>
`;
